subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,721 @@
+{
+	struct usbd_hid_descriptor *desc = c_nd->data->desc;
+
+	return desc->in_ep.bEndpointAddress;
+}
+
+static inline uint8_t hid_get_out_ep(struct usbd_class_node *const c_nd)
+{
+	struct usbd_hid_descriptor *desc = c_nd->data->desc;
+
+	return desc->out_ep.bEndpointAddress;
+}
+
+static int usbd_hid_request(struct usbd_class_node *const c_nd,
+			    struct net_buf *const buf, const int err)
+{
+	struct usbd_contex *uds_ctx = c_nd->data->uds_ctx;
+	const struct device *dev = c_nd->data->priv;
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	struct udc_buf_info *bi;
+
+	bi = udc_get_buf_info(buf);
+
+	if (bi->ep == hid_get_out_ep(c_nd)) {
+		if (ops->output_report != NULL) {
+			if (err == 0) {
+				ops->output_report(dev, buf->len, buf->data);
+			}
+
+			k_work_submit(&ddata->output_work);
+		}
+	}
+
+	if (bi->ep == hid_get_in_ep(c_nd)) {
+		if (ops->input_report_done != NULL) {
+			ops->input_report_done(dev);
+		} else {
+			k_sem_give(&ddata->in_sem);
+		}
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static int handle_set_idle(const struct device *dev,
+			   const struct usb_setup_packet *const setup)
+{
+	const uint32_t duration = HID_GET_IDLE_DURATION(setup->wValue);
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		ddata->idle_rate = duration;
+	}
+
+	LOG_DBG("Set Idle, Report ID %u Duration %u", id, duration);
+
+	if (ops->set_idle != NULL) {
+		ops->set_idle(dev, id, duration * 4UL);
+	} else {
+		errno = -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int handle_get_idle(const struct device *dev,
+			   const struct usb_setup_packet *const setup,
+			   struct net_buf *const buf)
+{
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	uint32_t duration;
+
+	if (setup->wLength != 1U) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (id != 0U && ops->get_idle == NULL) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		duration = ddata->idle_rate;
+	} else {
+		duration = ops->get_idle(dev, id) / 4UL;
+	}
+
+	LOG_DBG("Get Idle, Report ID %u Durationi %u", id, duration);
+	net_buf_add_u8(buf, duration);
+
+	return 0;
+}
+
+static int handle_set_report(const struct device *dev,
+			     const struct usb_setup_packet *const setup,
+			     const struct net_buf *const buf)
+{
+	const uint8_t type = HID_GET_REPORT_TYPE(setup->wValue);
+	const uint8_t id = HID_GET_REPORT_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (ops->set_report == NULL) {
+		errno = -ENOTSUP;
+		LOG_ERR("Set Report not supported");
+		return 0;
+	}
+
+	switch (type) {
+	case HID_REPORT_TYPE_INPUT:
```

Comment:
Input report type cannot be set, this should be an error case.

Reply:
Maybe I missed another ECN, but from the USB HID spec, both Get Report and Set Report are valid for all three report types. 7.2.1 "This request is useful at initialization time for absolute items and for determining the state of feature items. This request is not intended to be used for polling the device state on a regular basis." and 7.2.2 "The Set_Report request allows the host to send a report to the device, possibly setting the state of input, output, or feature controls."

Reply:
The same document talks about Input, Output and Feature reports in a different light in the Remarks at 6.2.2.5:
> Input items define input reports accessible via the Control pipe with a Get_Report (Input) request.

> Output items make Output reports accessible via the Control pipe with a Set_Report (Output) command.

> Output type reports can optionally be sent via an Interrupt Out pipe.

> Like Output items, Feature items make up Feature Reports accessible via the Control pipe with the Get_Report (Feature) and Set_Report (Feature) requests.

At the same time, this document is mixing these terms (Input, Output, Feature) between a category on the protocol level versus a category on the device - human interface level, the latter being a lot more permissive. All in all I think you are reading it right and sticking to the specification it should be allowed - even if it's very unlikely that any host OS ever does this. Maybe it would be worth it to try with a rawhid device and check if hidapi allows doing this.

Reply:
I think this is incorrect interpretation. The HID specification does explicitly mention `Set_Report(Output)` and `Set_Report(Feature)` while it never mentions an input version.

`6.2.2.5 Input, Output, and Feature Items` states:
> * An Input item describes information about the data provided by one or more physical controls. An application can use this information to interpret the data provided by the device. All data fields defined in a single item share an identical data format.
> * The Output item is used to define an output data field in a report. This item is similar to an Input item except it describes data sent to the device—for example, LED states.
> * Feature items describe device configuration information that can be sent to the device.

and later in the Remarks:
> * Input items define input reports accessible via the Control pipe with a Get_Report (Input) request.

> * Output items make Output reports accessible via the Control pipe with a Set_Report (Output) command.

> * Feature items may be attributes of other items. For example, an Origin Reset Feature may apply to one or more position Input items. Like Output items, Feature items make up Feature Reports accessible via the Control pipe with the Get_Report (Feature) and Set_Report (Feature) requests.

Reply:
Additionally HID Usage Tables Terms and Abbreviations define Application as:
> A software program that consumes the data generated by the HID device Input reports, or that controls the HID device through Feature or Output reports. Applications can be games or other programs used by end users or system software components.

Therefore it explicitly mentions that Input is from device to host, and only Feature or Output reports can control the device (from host to device). This matches the bullet points in HID Specification 6.2.2.5 Input, Output, and Feature Items.

Reply:
This is not about how reports are intended to be used, and the current implementation does not force anyone to use all the parameters in these callbacks. The spec explicitly mentions all three report types in the description of Get Report and Set Report. Set Report is even more specific, see 7.2.2,
"
```
A device might choose to ignore input Set_Report requests as meaningless.
Alternatively these reports could be used to reset the origin of a control (that
is, current position should report zero). The effect of sent reports will also
depend on whether the recipient controls are absolute or relative.
```
"